app/services/complete_moab_handler.rb

```diff
@@ -203,11 +204,14 @@ def create_db_objects(status, checksums_validated = false)
 
     # TODO: remove tests' dependence on 2 "create!" calls, use single built-in AR transactionality
     transaction_ok = with_active_record_transaction_and_rescue do
-      PreservedObject
-        .find_or_create_by!(druid: druid) do |po|
-          po.current_version = incoming_version
-          po.preservation_policy_id = ppid
-        end.complete_moabs.create!(cm_attrs)
+      this_po = PreservedObject
+                .find_or_create_by!(druid: druid) do |po|
+                  po.current_version = incoming_version
+                  po.preservation_policy_id = ppid
+                end
+      this_cm = this_po.complete_moabs.create!(cm_attrs)
+      # add to join table
+      PreservedObjectsPrimaryMoab.create!(preserved_object: this_po, complete_moab: this_cm)
```

Comment:
i think you want something like an `if this_po.complete_moabs.count == 1` guard clause, so that you're not accidentally marking a second moab primary if one gets picked up for the same druid/`PreservedObject` on a different storage root (now that `CompleteMoabHandler` can differentiate moabs for the same druid by storage root).

would probably be good to add a test that exercised that behavior.

Reply:
another option, if you want to mark the moab primary whenever there's not already a primary (as opposed to only marking it primary when its the first moab for the druid):  you could do a `find_or_create_by!` on line 214 instead of a `create!`.  that should be no less safe than the guard clause approach, since this entire section is wrapped in an explicit transaction, and already uses `find_or_create_by!` (which is not itself transactional).

i have a very slight preference for only making the moab the primary if its the first moab for the druid, because that seems a bit more deterministic, and a bit easier to test well.  but that's a _very_ slight preference, and i'm open to `find_or_create_by!` (or other approaches i haven't thought of).

Reply:
I actually checked the code and added the comment that the method is ONLY called when the complete_moab does not exist.  AND the db constraint will error if the record is already there.